### PR TITLE
Reworked decorations debounce delay with special exception for scrolling

### PR DESCRIPTION
**Pull request by AndreasArvidsson**
Removed single setting for debounce delay and replaced with two different default values. 
60Hz for all decorations except visible ranges at 25Hz to improve performance and remove annoyance when scrolling.

### CODE REVIEW

File: src/core/HatAllocator.ts [hunk 1]
```diff
@@ -37,15 +39,27 @@ export class HatAllocator {
...
+      // An Event which fires when the selection in an editor has changed.
       vscode.window.onDidChangeTextEditorSelection(
-        this.addDecorationsDebounced
+        this.addDecorationsDebouncedDefault
       ),
-      vscode.workspace.onDidChangeTextDocument(this.addDecorationsDebounced)
+      // An Event which fires when the visible ranges of an editor has changed.
+      vscode.window.onDidChangeTextEditorVisibleRanges(() =>
+        // Increase debounced delay to stop the decorations moving when scrolling. 25Hz
+        this.addDecorationsDebounced(40)
```
Comment (pokey) [inline src/core/HatAllocator.ts:61]: Move to constant at top?

File: src/core/HatAllocator.ts [hunk 2]
```diff
@@ -66,25 +80,25 @@ export class HatAllocator {
     }
   }
 
-  addDecorationsDebounced() {
+  // Use default debounced delay. 60Hz
+  addDecorationsDebouncedDefault() {
+    this.addDecorationsDebounced(16);
```
Comment (pokey) [inline src/core/HatAllocator.ts:85]: Move to constant at top?

File: src/core/HatAllocator.ts [hunk 3]
```diff
@@ -37,15 +39,27 @@ export class HatAllocator {
         this.toggleDecorations
       ),
 
-      vscode.window.onDidChangeTextEditorVisibleRanges(
-        this.addDecorationsDebounced
+      // An Event which fires when the active editor has changed. Note that the event also fires when the active editor changes to undefined.
```
Comment (pokey) [inline src/core/HatAllocator.ts:42]: Liking all the comments here; helpful

**Review by pokey — COMMENTED** [review]
Looks good but I'd like to have a quick play with it before we merge

File: src/core/HatAllocator.ts (hunk 1 above)
Comment (AndreasArvidsson) [reply src/core/HatAllocator.ts:61]: Idea is that if more events want to deviate from the default we can add a different delay for each single event. A constant at the top would make more sense if we actually reused it multiple times.

File: src/core/HatAllocator.ts (hunk 2 above)
Comment (AndreasArvidsson) [reply src/core/HatAllocator.ts:85]: I mean it's only every used at one place. What would the constant improve?

File: src/core/HatAllocator.ts (hunk 3 above)
Comment (AndreasArvidsson) [reply src/core/HatAllocator.ts:42]: I thought so too. Some of the names aren't that clear.

File: src/core/HatAllocator.ts (hunk 2 above)
Comment (pokey) [reply src/core/HatAllocator.ts:85]: I'm not a fan of magic numbers in code, but I guess I don't feel strongly here as you have a comment right above it explaining the magic number

File: src/core/HatAllocator.ts (hunk 1 above)
Comment (pokey) [reply src/core/HatAllocator.ts:61]: As below, I have a philosophical objection but I think it's prob fine here as well

**Comment by pokey** [comment]
Ok just wired it up.  Def nice to get that 16ms thing, but I think 40ms is a bit fast for scroll, as there are a bunch of refreshes as my momentum scrolling slows at the end.  Maybe we could bump it up a little bit, or make it a setting?

File: src/core/HatAllocator.ts (hunk 2 above)
Comment (AndreasArvidsson) [reply src/core/HatAllocator.ts:85]: I agree that unnamed and undocument that numbers is not a good thing. In this case I think we're fine due to the comment.

**Review by pokey — APPROVED** [review]

**Pull request merged**